soql-standalone-parser/src/main/scala/com/socrata/soql/parsing/RecursiveDescentParser.scala

```diff
@@ -0,0 +1,1521 @@
+    reader.first match {
+      case i: Identifier => ParseResult(reader.rest, (i.value, i.position))
+      case _ : SystemIdentifier => fail1(reader, ANonSystemIdentifier)
+      case _ => fail1(reader, AnIdentifier)
+    }
+
+  private def userIdentifier(reader: Reader): ParseResult[(ColumnName, Position)] = {
+    simpleIdentifier(reader).map { case (n, p) => (ColumnName(n), p) }
+  }
+
+  private def identifier(reader: Reader): ParseResult[(ColumnName, Position)] = {
+    reader.first match {
+      case i: Identifier =>
+        ParseResult(reader.rest, (ColumnName(i.value), i.position))
+      case si: SystemIdentifier =>
+        ParseResult(reader.rest, (ColumnName(si.value), si.position))
+      case _ =>
+        fail1(reader, AnIdentifier)
+    }
+  }
+
+  private def userStarSelections(reader: Reader, commaFirst: Boolean): ParseResult[Seq[StarSelection]] = {
+    // This is a little tricky, because we may or may not be parsing
+    // anything at all, so we need to keep track of when we've
+    // committed to a thing.  Basically we're parsing either
+    //     (COMMA userStarExpr)*
+    // if commaFirst is true and
+    //     (userStarExpr (COMMA userStarExpr)*)?
+    // if it's false, and we're only committed to completing a phrase
+    // at the points where we've actually seen a STAR token
+
+    val result = Vector.newBuilder[StarSelection]
+
+    @tailrec
+    def loop(reader: Reader, commaFirst: Boolean): Reader = {
+      val r =
+        if(commaFirst) {
+          reader.first match {
+            case COMMA() =>
+              reader.rest
+            case _ =>
+              // We require a comma but do not see one - revert.
+              reader.addAlternates(COMMASET)
+              return reader
+          }
+        } else {
+          reader
+        }
+
+      r.first match {
+        case ti: TableIdentifier =>
+          r.rest.first match {
+            case DOT() =>
+              r.rest.rest.first match {
+                case star@STAR() =>
+                  // We are now committed to parsing a user-star-selection
+                  val ParseResult(r2, exceptions) = selectExceptions(r.rest.rest.rest, userIdentifier)
+                  result += StarSelection(Some(intoQualifier(ti)), exceptions).positionedAt(star.position)
+                  loop(r2, commaFirst = true)
+                case _ =>
+                  // Whoops, not a star selection, revert
+                  reader.addAlternates(TABLEIDSTARSET)
+                  reader
+              }
+            case _ =>
+              reader.addAlternates(TABLEIDSTARSET)
+              reader
+          }
+        case star@STAR() =>
+          val ParseResult(r2, exceptions) = selectExceptions(r.rest, userIdentifier)
+          result += StarSelection(None, exceptions).positionedAt(star.position)
+          loop(r2, commaFirst = true)
+        case _ =>
+          reader.addAlternates(TABLEIDSTARSET)
```

Comment:
This pings weirdly for me, but it's the same as the systemStarSelection bit, so it's probably fine.

Mostly finding it odd that the error message will suggest a table-identifer-plus-star when... not finding one is perfectly okay.

Reply:
It _does_ make sense on a node level, and I have the same kind of "wait, this should never be relevant" confusion elsewhere, so I'm inclined to feel that this is a completeness "cover all cases, whether or not they can be walked into naturally because we might need to walk into them from a different direction someday" thing rather than a mistake.